port/posix/posix_transport_shm.c

```diff
@@ -1,301 +1,581 @@
+            /* Some other error */
+            ret = WH_ERROR_ABORTED;
         }
     }
-    else {
-        /* Something went wrong with unlink, we are unable to remove the
-         * existing shared memory object, bail */
-        fprintf(
-            stderr,
-            "Error: Shared memory object already exists, unable to delete.\n");
-        exit(EXIT_FAILURE);
+    return ret;
+}
+
+
+static int posixTransportShm_CreateMap(char* name, uint16_t req_size,
+        uint16_t resp_size, size_t dma_size, ptshmMapping* map)
+{
+    int ret = WH_ERROR_OK;
+    int fd = -1;
+
+    if (    (name == NULL) ||
+            (map == NULL) ) {
+        return WH_ERROR_BADARGS;
     }
 
-    /* Map the shared memory object */
-    ctx->shmBuf =
-        mmap(NULL, shmSize, PROT_READ | PROT_WRITE, MAP_SHARED, shmFd, 0);
-    if (ctx->shmBuf == MAP_FAILED) {
-        perror("mmap");
-        close(shmFd);
-        exit(EXIT_FAILURE);
+    /* Attempt to remove any existing object. */
+    (void)shm_unlink(name);
+    /* Create shared memory object and set the size */
+    fd = shm_open(name, O_CREAT | O_RDWR, PTSHM_CREATEMODE);
+    if (fd >= 0) {
+        /* Set the size of the shared memory object.
+         * Note this is the minimum size, as the OS may make it larger. */
+        size_t size = sizeof(*(map->header)) + req_size + resp_size + dma_size;
+        if (ftruncate(fd, size) == 0) {
+            /* Map the header and set the configuration */
+            ptshmHeader* header = (ptshmHeader*)mmap(NULL, sizeof(*header),
+                    PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+            if (header != MAP_FAILED) {
+                header->req_size = req_size;
+                header->resp_size = resp_size;
+                header->dma_size = dma_size;
+                header->creator_pid = getpid();
+                XMEMFENCE();
+                header->initialized = PTSHM_INITIALIZED_CREATOR;
+
+                /* Unmap the header and remap the full area */
+                (void)munmap((void*)header, sizeof(*header));
+                ret = posixTransportShm_Map(fd, size, map);
+            } else {
+                /* Problem mapping the header */
+                ret = WH_ERROR_ABORTED;
+            }
+        } else {
+            /* Problem setting the size. */
+            ret = WH_ERROR_ABORTED;
+        }
+        close(fd);
+    } else {
+        /* Problem creating the shared memory */
+        ret = WH_ERROR_ABORTED;
     }
-    header = (ShmHeader*)ctx->shmBuf;
+    return ret;
+}
 
-    /* Initialize or increment the reference count */
-    atomic_fetch_add(&header->refCount, 1);
 
-    /* allocate a shared memory transport context */
-    ctx->transportMemCtx = malloc(sizeof(whTransportMemContext));
-    if (ctx->transportMemCtx == NULL) {
-        return WH_ERROR_ABORTED;
+/** Custom functions */
+int posixTransportShm_IsConnected(posixTransportShmContext* ctx,
+        whCommConnected *out_connected)
```

Comment:
function should probably be static

Reply:
This is exposed to the user so they will know when the DMA region is usable by both sides.  Really, the server (creator) needs to understand this since the client (user) will be the last to map it.  The transport handles the connected state internally.